okhttp/src/main/java/okhttp3/internal/http1/Http1ExchangeCodec.java

```diff
@@ -13,530 +13,542 @@
-    private boolean closed;
+    @Throws(IOException::class)
+    override fun readResponseHeaders(expectContinue: Boolean): Response.Builder? {
+        if (state != STATE_OPEN_REQUEST_BODY && state != STATE_READ_RESPONSE_HEADERS) {
+            throw IllegalStateException("state: $state")
+        }
 
-    ChunkedSink() {
+        try {
+            val statusLine = StatusLine.parse(readHeaderLine())
+
+            val responseBuilder = Response.Builder()
+                    .protocol(statusLine.protocol)
+                    .code(statusLine.code)
+                    .message(statusLine.message)
+                    .headers(readHeaders())
+
+            if (expectContinue && statusLine.code == HTTP_CONTINUE) {
+                return null
+            } else if (statusLine.code == HTTP_CONTINUE) {
+                state = STATE_READ_RESPONSE_HEADERS
+                return responseBuilder
+            }
+
+            state = STATE_OPEN_RESPONSE_BODY
+            return responseBuilder
+        } catch (e: EOFException) {
+            // Provide more context if the server ends the stream before sending a response.
+            val address = if (realConnection != null) realConnection.route().address().url().redact() else "unknown"
+            throw IOException("unexpected end of stream on $address", e)
+        }
     }
 
-    @Override public Timeout timeout() {
-      return timeout;
+    @Throws(IOException::class)
+    private fun readHeaderLine(): String {
+        val line = source.readUtf8LineStrict(headerLimit)
+        headerLimit -= line.length.toLong()
+        return line
     }
 
-    @Override public void write(Buffer source, long byteCount) throws IOException {
-      if (closed) throw new IllegalStateException("closed");
-      if (byteCount == 0) return;
+    /** Reads headers or trailers.  */
+    @Throws(IOException::class)
+    private fun readHeaders(): Headers {
+        val headers = Headers.Builder()
+        // parse the result headers until the first blank line
+        var line = readHeaderLine()
+        while (line.isNotEmpty()) {
+            addHeaderLenient(headers, line)
+            line = readHeaderLine()
+        }
+        return headers.build()
+    }
 
-      sink.writeHexadecimalUnsignedLong(byteCount);
-      sink.writeUtf8("\r\n");
-      sink.write(source, byteCount);
-      sink.writeUtf8("\r\n");
+    private fun newChunkedSink(): Sink {
+        if (state != STATE_OPEN_REQUEST_BODY) throw IllegalStateException("state: $state")
+        state = STATE_WRITING_REQUEST_BODY
+        return ChunkedSink()
     }
 
-    @Override public synchronized void flush() throws IOException {
-      if (closed) return; // Don't throw; this stream might have been closed on the caller's behalf.
-      sink.flush();
+    private fun newKnownLengthSink(): Sink {
+        if (state != STATE_OPEN_REQUEST_BODY) throw IllegalStateException("state: $state")
```

Comment:
`check(...)` here and elsewhere

Reply:
Think we need some inlinable conversion for this, made a trivial mistake while flipping the bits.  Easier to retain, inline, then simplify with Intellij.

Reply:
I add an empty else clause, then use “invert if” and I have all the missing pieces I need.